src/predictor/array_tree_layout.h

```diff
@@ -0,0 +1,273 @@
+    static_assert(std::is_same_v<RegTree, TreeType>);
+    return tree[nidx].LeftChild() + 1;
+  }
+
+  inline static bst_node_t RightChild(const MultiTargetTree& tree, bst_node_t nidx) {
+    static_assert(std::is_same_v<MultiTargetTree, TreeType>);
+    return tree.RightChild(nidx);
+  }
+
+ /**
+ * @brief Traverse the top levels of original tree and fill internal arrays
+ * 
+ * \tparam TreeType The type of the origianl tree (RegTree or MultiTargetTree)
+ *
+ * \tparam depth the tree level being processing
+ *
+ * \param tree the original tree
+ * 
+ * \param cats matrix of categorical splits
+ * 
+ * \param nidx_array node idx in the array layout
+ * 
+ * \param nidx node idx in the original tree
+ * 
+ */
+  template <int depth = 0>
+  void inline Populate(const TreeType& tree, RegTree::CategoricalSplitMatrix const &cats,
+                       bst_node_t nidx_array = 0, bst_node_t nidx = 0) {
+    if constexpr (depth == kNumDeepLevels + 1) {
+      return;
+    } else if constexpr (depth == kNumDeepLevels) {
+        /* We save the node index in the origianl tree to able to continue processing
+         * for nodes not egligable for array layout optimisation. 
+         */
+        nidx_in_tree_[nidx_array - kNodesCount] = nidx;
+    } else {
+      if (IsLeaf(tree, nidx)) {
+        split_index_[nidx_array]  = 0;
+
+        /* 
+         * if the tree is not fully populated, we can reduce transfering costs.
+         * the values for unpopulated part of the tree are set in a way to guarantie
+         * that a moove will always done in "right" direction.
+         * here we exploiting that comparison with nan always results to false.
+         */
+        if constexpr (any_missing) default_left_[nidx_array] = 0;
+        if constexpr (has_categorical) is_cat_[nidx_array] = 0;
+        split_cond_[nidx_array]   = std::numeric_limits<float>::quiet_NaN();
+
+        Populate<depth + 1>(tree, cats, 2 * nidx_array + 2, nidx);
+      } else {
+        if constexpr (any_missing) default_left_[nidx_array] = DefaultLeft(tree, nidx);
+        if constexpr (has_categorical) {
+          is_cat_[nidx_array] = common::IsCat(cats.split_type, nidx);
+          if (is_cat_[nidx_array]) {
+            cat_segment_[nidx_array] = cats.categories.subspan(cats.node_ptr[nidx].beg,
+                                                               cats.node_ptr[nidx].size);
+          }
+        }
+
+        split_index_[nidx_array]  = SplitIndex(tree, nidx);
+        split_cond_[nidx_array]   = SplitCond(tree, nidx);
+
+        /*
+         * LeftChild is used to find if the node is leaf, so it is a valid value,
+         * howerwer RightChild can be invalid in some exotic case.
+         * The tree with invalid right-child can be correctly processed by a classical method,
+         * if the split conditions are propper.
+         * But for array layout invalid RightChild, even unreachable, will lead to memory corruption.
+         * Add check to prevent it.
```

Comment:
```suggestion
         * LeftChild is used to determine if a node is a leaf, so it is always a valid value.
         * However, RightChild can be invalid in some exotic cases.
         * A tree with an invalid RightChild can still be correctly processed using classical methods
         * if the split conditions are correct.
         * However, in an array layout, an invalid RightChild, even if unreachable, can lead to memory corruption.
         * A check should be added to prevent this.
```